File: Java/src/NapakalakiGame/Player.java
/*
 * Autores:
 * Guillermo Montes Martos
 * Manuel Ruiz López
 *
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;
import GUI.Dice;

public class Player {
    
    static final int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead = true;
    private boolean canISteal = true;
    private ArrayList<Treasure> hiddenTreasures = new ArrayList();
    private ArrayList<Treasure> visibleTreasures = new ArrayList();
    protected Player enemy;
    private BadConsequence pendingBadConsequence;
 
    
    public Player (String aname){
        name = aname;
        level = 1;
        pendingBadConsequence = new NumericBadConsequence("", 0, 0, 0);
    }
    
    public Player (Player p){
        this.name = p.name;
        this.level = p.level;
        this.dead = p.dead;
        this.canISteal = p.canISteal;
        this.hiddenTreasures = new ArrayList(p.hiddenTreasures);
        this.visibleTreasures = new ArrayList(p.visibleTreasures);
        this.enemy = p.enemy;
        this.pendingBadConsequence = p.pendingBadConsequence;
    }
    
    public String getName (){
        return name;
    }
    
    private void bringToLife(){
        dead = false;
    }
    
    public int getCombatLevel(){
        int suma = level;
        for (Treasure elemento:visibleTreasures)
            suma += elemento.getBonus();
        
        return suma;
    }
    
    private void incrementLevels(int i){
        level += i;
        
        if (level > MAXLEVEL)
            level = MAXLEVEL;
    } 
    
    private void decrementLevels(int i){
        level -= i;
        
        if (level < 1)
            level = 1;
    }
    
    public BadConsequence getPendingBadConsequence(){
        return pendingBadConsequence;
    }
    
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence = b;
    }
    
    private void applyPrize (Monster m){
        incrementLevels(m.getLevelsGained());
        int nTreasures = m.getTreasuresGained();
        if(nTreasures>0) {
            CardDealer dealer = CardDealer.getInstance();
            for(int i=0; i<nTreasures; i++)
                hiddenTreasures.add(dealer.nextTreasure());
        }
    }
    
    private void applyBadConsequence (Monster m){
        BadConsequence bc = m.getBadConsequence();
        decrementLevels(bc.getLevels());
        setPendingBadConsequence(bc.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures));
    }
    
    
    private boolean canMakeTreasureVisible(Treasure t){
        boolean condition = true;
        
        int i = 0;
        switch (t.getType()){
            case BOTHHANDS:
                while (i<visibleTreasures.size() && condition){
                    if ( (visibleTreasures.get(i).getType() == TreasureKind.ONEHAND) ||
                        (visibleTreasures.get(i).getType() == TreasureKind.BOTHHANDS) )
                        condition = false;
                    else
                        i++;
                }
                
                break;
                
                
            case ONEHAND:
                boolean one_hand = false;
                
                while(i<visibleTreasures.size() && condition){
                    if (visibleTreasures.get(i).getType() == TreasureKind.BOTHHANDS)
                        condition = false;
                    else if (visibleTreasures.get(i).getType() == TreasureKind.ONEHAND)
                        if (one_hand)
                            condition = false;
                        else{
                            one_hand = true;
                            i++;
                        }
                    else
                        i++;
                }
                
                break;
                
                
            default:
                while (i<visibleTreasures.size() && condition){
                    if (visibleTreasures.get(i).getType() == t.getType())
                        condition = false;
                    else 
                        i++;
                }
                
                break;
                
        }
        
        return condition;
    }
    
    private int howManyVisibleTreasures(TreasureKind tKind){
        int contador = 0;
        for (Treasure elemento:visibleTreasures)
            if (elemento.getType() == tKind)
                contador++;
        
        return contador;
    }
    
    private void dieIfNoTreasures(){
        if (hiddenTreasures.isEmpty() && visibleTreasures.isEmpty())
            dead = true;
    }
    
    public boolean isDead(){
        return dead;
    }

    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }

    public ArrayList<Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
    
    
    public CombatResult combat (Monster m){
        CombatResult result;
        
        if(getCombatLevel() > getOponentLevel(m)) {
            applyPrize(m);
            result = (level>=MAXLEVEL)? CombatResult.WINGAME : CombatResult.WIN;
        }
        else {
            applyBadConsequence(m);
            result = shouldConvert()? CombatResult.LOSEANDCONVERT : CombatResult.LOSE;
        }
        
        return result;
    }
    
    public void makeTreasureVisible(Treasure t){
        if(canMakeTreasureVisible(t)) {
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure (Treasure t){
        visibleTreasures.remove(t);
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractVisibleTreasure(t);
        dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure (Treasure t){
        hiddenTreasures.remove(t);
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractHiddenTreasure(t);
        dieIfNoTreasures();
    }
    
    public boolean validState (){        
        return (hiddenTreasures.size()<5 && (pendingBadConsequence==null || pendingBadConsequence.isEmpty()));
    }
    
    public void initTreasures(String nombre){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        int number;
        
        bringToLife();
        hiddenTreasures.add(dealer.nextTreasure());
        number = dice.nextNumber(nombre+", en un momento estarás vivo.", "Un número grande te dará más tesoros.");
        if(number>1) {
            hiddenTreasures.add(dealer.nextTreasure());
            if(number==6)
                hiddenTreasures.add(dealer.nextTreasure());
        }
    }
    
    public int getLevels(){
        return level;
    }
    
    public Treasure stealTreasure(){
        Treasure treasure = null;
        
        if(canISteal && enemy.canYouGiveMeATreasure())
        {
            treasure = enemy.giveMeATreasure();
            hiddenTreasures.add(treasure);
            haveStolen();
        }
        
        return treasure;
    }
       
    public void setEnemy (Player enemy){
        this.enemy = enemy;
    }
    
    protected Treasure giveMeATreasure (){
        Random n = new Random();
        return hiddenTreasures.remove(n.nextInt(hiddenTreasures.size()));
    }
    
    public boolean canISteal (){
        return canISteal;
    }
    
    protected boolean canYouGiveMeATreasure(){
        return (!hiddenTreasures.isEmpty());
    }
    
    private void haveStolen(){
        canISteal = false;
    }
    
    public void discardAllTreasures(){
        while(!visibleTreasures.isEmpty())
            discardVisibleTreasure(visibleTreasures.get(0));
        while(!hiddenTreasures.isEmpty())
            discardHiddenTreasure(hiddenTreasures.get(0));
    }
    
    @Override
    public String toString() {
        return name;
    }
   
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        return Dice.getInstance().nextNumber("Lo siento, has perdido el combate.", 
                             "Si sacas un 6, te convertirás en jugador sectario.") == 6;           
    }
    
    public Player getEnemy(){
        return enemy;
    }
      
    
}

File: Ruby/lib/Player.rb
#encoding: utf-8

# Autores:
# 
#     Guillermo Montes Martos
#     Manuel Ruiz López
#

require_relative 'Treasure.rb'
require_relative 'BadConsequence.rb'
require_relative 'Dice.rb'
require_relative 'TreasureKind.rb'

module NapakalakiGame
  class Player
    
    #class << self
    #end
    
    attr_reader :name
    attr_reader :level
    attr_reader :dead
    attr_reader :visibleTreasures
    attr_reader :hiddenTreasures
    attr_reader :canISteal
    
    attr_writer :enemy
    attr_writer :pendingBadConsequence
    
    # No hace falta poner attr_reader para @@MAXLEVEL porque el funcionamiento  
    # de las variables de clase en ruby es diferente a las variables de instancia.
    # Para consultar su valor: Player.class_variable_get(:@@MAXLEVEL)
    @@MAXLEVEL = 10

    
    def initialize(name)
      @name = name
      @level = 1
      @dead = true
      @canISteal = true
      @hiddenTreasures = Array.new
      @visibleTreasures = Array.new
    end
    
    def self.crearCopia(p)
      new(p.name)
      @level = p.level
      @dead = p.dead
      @canISteal = p.canISteal
      @hiddenTreasures = p.hiddenTreasures
      @visibleTreasures = p.visibleTreasures
    end
    
    protected
    def getOponentLevel(m)
      m.combatLevel
    end
    
    def shouldConvert()
      dice = Dice.instance
      dice.nextNumber == 1
    end
    
    def getCombatLevel()
      combatlevel = @level
      
      for i in @visibleTreasures
        combatlevel += i.bonus
      end
       
      combatlevel
    end
    
    private
    def bringToLife()
      @dead = false
    end
    
    def incrementLevels(l)
      @level += l
      
      if(@level > @@MAXLEVEL)
        @level = @@MAXLEVEL
      end
    end
    
    def decrementLevels(l)
      @level -= l
      
      if(@level < 1)
        @level = 1
      end
    end
    
    def applyPrize(m)
      incrementLevels(m.getLevelsGained)
      
      if (m.getTreasuresGained > 0)
        for i in 0..m.getTreasuresGained
          @hiddenTreasures << CardDealer.instance.nexTreasure
        end
      end
      
    end
    
    def applyBadConsequence(m)
      bc = m.badConsequence
      decrementLevels(bc.levels)
      @pendingBadConsequence = bc.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
    end
    
    def canMakeTreasureVisible(t)
      condicion = true
      
      case t.type
      when TreasureKind::BOTHHANDS
        i = 0
        
        while (condicion && i<@visibleTreasures.length) do
          if (@visibleTreasures[i].type==TreasureKind::BOTHHANDS ||
              @visibleTreasures[i].type==TreasureKind::ONEHAND)
            condicion = false
          else
            i+=1
          end
        end
      
      when TreasureKind::ONEHAND
        i = 0
        tiene_onehand = false
        
        while (condicion && i<@visibleTreasures.length) do
          if (@visibleTreasures[i].type==TreasureKind::BOTHHANDS)
            condicion = false
          elsif (@visibleTreasures[i].type==TreasureKind::ONEHAND)
            if(tiene_onehand)
              condicion = false
            else
              tiene_onehand = true
              i += 1
            end
          else
            i+=1
          end
        end
        
      else
        i = 0
        
        while (condicion && i<@visibleTreasures.length) do
          if (t.type==@visibleTreasures[i].type)
            condicion = false
          else
            i+=1
          end
        end
        
      end
              
      condicion
    end

    def howManyVisibleTreasures(tKind)
      contador = 0
      
      for i in @visibleTreasures
        contador += 1 if (i.type == tKind)
      end
      
      contador
    end
    
    def dieIfNoTreasures()
      @dead = true if(@visibleTreasures.empty? && @hiddenTreasures.empty?)
    end
    
    public    
    def combat(m)
      if getCombatLevel() > getOponentLevel(m)
        
        applyPrize(m)
        
        if @level >= @@MAXLEVEL
          CombatResult::WINGAME
        else
          CombatResult::WIN
        end
        
      else
        applyBadConsequence(m)
        
        if(shouldConvert)
          CombatResult::LOSEANDCONVERT
        else
          CombatResult::LOSE
        end
        
      end
    end
    
    def makeTreasureVisible(t)
      if canMakeTreasureVisible(t)
        @visibleTreasures << t
        @hiddenTreasures.delete_at(@hiddenTreasures.index(t))
      end
    end
    
    def discardVisibleTreasure(t)
      @visibleTreasures.delete_at(@visibleTreasures.index(t))
      if (@pendingBadConsequence != nil && !@pendingBadConsequence.isEmpty)
        @pendingBadConsequence.substractVisibleTreasure(t)
      end
      dieIfNoTreasures
    end
    
    def discardHiddenTreasure(t)
      @hiddenTreasures.delete_at(@hiddenTreasures.index(t))
      if (@pendingBadConsequence != nil && !@pendingBadConsequence.isEmpty)
        @pendingBadConsequence.substractHiddenTreasure(t)
      end
      dieIfNoTreasures
    end
    
    def validState()
      @hiddenTreasures.length < 5 && (@pendingBadConsequence == nil || @pendingBadConsequence.isEmpty)
    end
    
    def initTreasures
      
      bringToLife
      @hiddenTreasures << CardDealer.instance.nexTreasure
      dice = Dice.instance
      number = dice.nextNumber
      
      if number > 1
        @hiddenTreasures << CardDealer.instance.nexTreasure
        if number == 6
          @hiddenTreasures << CardDealer.instance.nexTreasure
        end
      end
      
    end
    
    def stealTreasure()
      treasure = nil

      if @canISteal && @enemy.canYouGiveMeATreasure()
        treasure = @enemy.giveMeATreasure
        @hiddenTreasures << treasure
        haveStolen()
      end
      treasure
    end
    
    protected
    def giveMeATreasure()
      @hiddenTreasures.delete_at(rand(@hiddenTreasures.length))
    end
    
    def canYouGiveMeATreasure()
      !@hiddenTreasures.empty?
    end
    
    private
    def haveStolen()
      @canISteal = false
    end
    
    public
    def discardAllTreasures()
      # Mientras haya elementos en el vector, se borra el primero de ellos
      while(!@visibleTreasures.empty?)
        discardVisibleTreasure(@visibleTreasures[0])
      end
      
      while(!@hiddenTreasures.empty?)
        discardHiddenTreasure(@hiddenTreasures[0])
      end
    end
    
    def to_s()
      "#{@name} (#{@level})"
    end
    
  end
end
